pkg/semtech-loramac/contrib/semtech_loramac.c

```diff
@@ -0,0 +1,690 @@
+            msg.type = MSG_TYPE_RX_TIMEOUT;
+
+            if (msg_send(&msg, _mac_pid) <= 0) {
+                DEBUG("[semtech-loramac] RX timeout, possibly lost interrupt.\n");
+            }
+            break;
+
+        case NETDEV_EVENT_CRC_ERROR:
+            DEBUG("[semtech-loramac] RX CRC error\n");
+            radio_events.RxError();
+            break;
+
+        case NETDEV_EVENT_FHSS_CHANGE_CHANNEL:
+            DEBUG("[semtech-loramac] FHSS channel change\n");
+            radio_events.FhssChangeChannel(((sx127x_t*)dev)->_internal.last_channel);
+            break;
+
+        case NETDEV_EVENT_CAD_DONE:
+            DEBUG("[DEBUG] test: CAD done\n");
+            radio_events.CadDone(((sx127x_t*)dev)->_internal.is_last_cad_success);
+            break;
+
+        default:
+            DEBUG("[semtech-loramac] unexpected netdev event received: %d\n",
+                  event);
+            break;
+    }
+}
+
+void *_event_loop(void *arg)
+{
+    static msg_t _msg_q[LORAMAC_MSG_QUEUE];
+    msg_init_queue(_msg_q, LORAMAC_MSG_QUEUE);
+    LoRaMacPrimitives_t primitives;
+    LoRaMacCallback_t callbacks;
+
+    _init_loramac(&primitives, &callbacks);
+    semtech_loramac_set_dr(LORAMAC_DEFAULT_DR);
+    semtech_loramac_set_adr(LORAMAC_DEFAULT_ADR);
+    semtech_loramac_set_public_network(LORAMAC_DEFAULT_PUBLIC_NETWORK);
+
+    while (1) {
+        msg_t msg;
+        msg_receive(&msg);
+        switch(msg.type) {
+            case MSG_TYPE_ISR:
+            {
+                netdev_t *dev = msg.content.ptr;
+                dev->driver->isr(dev);
+            }
+                break;
+
+            case MSG_TYPE_RX_TIMEOUT:
+                DEBUG("[semtech-loramac] RX timer timeout\n");
+                radio_events.RxTimeout();
+                break;
+
+            case MSG_TYPE_TX_TIMEOUT:
+                DEBUG("[semtech-loramac] TX timer timeout\n");
+                radio_events.TxTimeout();
+                break;
+
+            case MSG_TYPE_MAC_TIMEOUT:
+            {
+                DEBUG("[semtech-loramac] MAC timer timeout\n");
+                void (*callback)(void) = msg.content.ptr;
+                callback();
+            }
+                break;
+            case MSG_TYPE_LORAMAC_CMD:
+            {
+                DEBUG("[semtech-loramac] loramac cmd\n");
+                semtech_loramac_call_t *call = msg.content.ptr;
+                call->func(call->arg);
+            }
+                break;
+
+            default:
+                DEBUG("[semtech-loramac] Unexpected msg type '%04x'\n", msg.type);
+                break;
+        }
+    }
+}
+
+int semtech_loramac_init(sx127x_t *dev)
+{
+    dev->netdev.driver = &sx127x_driver;
+    dev->netdev.event_callback = _event_cb;
+
+    _main_pid = thread_getpid();
+    _mac_pid = thread_create(stack, sizeof(stack), THREAD_PRIORITY_MAIN - 1,
+                             THREAD_CREATE_STACKTEST, _event_loop, NULL,
+                             "recv_thread");
```

Comment:
IMHO we should do aim for one of the following:

1. have LoRAMAC be a netdev layer
2. have LoRAMAC be a `gnrc_netif` thread

From the current code base the latter would be easier to port to and my preferred method (since layered netdev is only WIP at the moment). In any case LoRAMAC should not be a completely separate thread from `gnrc_netif`.